plugins/inputs/clickhouse/clickhouse.go

```diff
@@ -0,0 +1,355 @@
+type ClickHouse struct {
+	Servers            []string `toml:"servers"`
+	AutoDiscovery      bool     `toml:"auto_discovery"`
+	ClusterInclude     []string `toml:"cluster_include"`
+	ClusterExclude     []string `toml:"cluster_exclude"`
+	Timeout            int      `toml:"timeout"`
+	InsecureSkipVerify bool     `toml:"http_tls_insecure_skip_verify"`
+	client             http.Client
+}
+
+// SampleConfig returns the sample config
+func (*ClickHouse) SampleConfig() string {
+	return sampleConfig
+}
+
+// Description return plugin description
+func (*ClickHouse) Description() string {
+	return "Read metrics from one or many ClickHouse servers"
+}
+
+// Start ClickHouse input service
+func (ch *ClickHouse) Start(telegraf.Accumulator) error {
+	timeout := defaultTimeout
+	if ch.Timeout != 0 {
+		timeout = time.Duration(ch.Timeout) * time.Second
+	}
+	ch.client = http.Client{
+		Timeout: timeout,
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: ch.InsecureSkipVerify,
+			},
+		},
+	}
+	return nil
+}
+
+// Gather collect data from ClickHouse server
+func (ch *ClickHouse) Gather(acc telegraf.Accumulator) (err error) {
+	var (
+		connects []connect
+		exists   = func(host string) bool {
+			for _, c := range connects {
+				if c.Hostname == host {
+					return true
+				}
+			}
+			return false
+		}
+	)
+
+	for _, server := range ch.Servers {
+		u, err := url.Parse(server)
+		if err != nil {
+			return err
+		}
+		switch {
+		case ch.AutoDiscovery:
+			var conns []connect
+			if err := ch.execQuery(u, "SELECT cluster, shard_num, host_name FROM system.clusters "+ch.clusterIncludeExcludeFilter(), &conns); err != nil {
+				acc.AddError(err)
+				continue
+			}
+			for _, c := range conns {
+				if !exists(c.Hostname) {
+					c.url = &url.URL{
+						Scheme: u.Scheme,
+						Host:   net.JoinHostPort(c.Hostname, u.Port()),
+						User:   u.User,
+					}
+					connects = append(connects, c)
+				}
+			}
+		default:
+			connects = append(connects, connect{
+				url: u,
+			})
+		}
+	}
+
+	for _, conn := range connects {
+		if err := ch.tables(acc, &conn); err != nil {
+			acc.AddError(err)
+		}
+		for metric := range commonMetrics {
+			if err := ch.commonMetrics(acc, &conn, metric); err != nil {
+				acc.AddError(err)
+			}
+		}
+	}
+	return nil
+}
+
+func (ch *ClickHouse) clusterIncludeExcludeFilter() string {
+	if len(ch.ClusterInclude) == 0 && len(ch.ClusterExclude) == 0 {
+		return ""
+	}
+	var (
+		escape = func(in string) string {
+			return "'" + strings.NewReplacer(`\`, `\\`, `'`, `\'`).Replace(in) + "'"
+		}
+		makeFilter = func(expr string, args []string) string {
+			in := make([]string, 0, len(args))
+			for _, v := range args {
+				in = append(in, escape(v))
+			}
+			return "cluster " + expr + " (" + strings.Join(in, ", ") + ")"
+		}
+	)
+
+	if len(ch.ClusterInclude) != 0 {
+		return "WHERE " + makeFilter("IN", ch.ClusterInclude)
+	}
+
+	return "WHERE " + makeFilter("NOT IN", ch.ClusterExclude)
```

Comment:
it seems like this code would make it so that you can't use both filters at the same time. shouldn't be too hard to adjust to support both

Reply:
fixed on https://github.com/influxdata/telegraf/pull/6441/commits/353c73b06df0fc8584c0da13a72c178060c5215c